scripts/context_retrieval/build_database.py

```diff
@@ -0,0 +1,652 @@
+"""
+Downloads and persists SQuAD training data. Builds and persists LlamaIndex database. Computes OpenAI
+embeddings for query and database data.
+"""
+
+import json
+import logging
+import os
+import sys
+import urllib
+from typing import Any, List, Tuple, TypedDict
+
+import numpy as np
+import openai  # type: ignore
+import pandas as pd
+import tiktoken  # type: ignore
+import tqdm  # type: ignore
+from llama_index import GPTVectorStoreIndex  # type: ignore
+from llama_index.data_structs.node import DocumentRelationship, Node  # type: ignore
+from typing_extensions import TypeAlias
+
+EmbeddingVector: TypeAlias = Any
+
+
+class DatabaseData(TypedDict):
+    article_index: List[int]
+    paragraph_index: List[int]
+    granular_subject: List[str]
+    broad_subject: List[str]
+    text: List[str]
+
+
+class QueryData(TypedDict):
+    id: List[str]
+    granular_subject: List[str]
+    broad_subject: List[str]
+    paragraph_index: List[int]
+    text: List[str]
+    is_answerable: List[bool]
```

Comment:
I would take advantage of being able to annotate these classes w/ explanations of the data